File: src/domain/matching.test.ts
import { describe, expect, it } from 'vitest';
import type { Checklist, ChecklistEntry } from '../types';
import { checklist } from '../data/checklist';
import {
  bestHighConfidenceConfusionMatchFromText,
  bestHighConfidenceExactAliasMatchFromText,
  bestMatchFromText,
  extractCodes,
  matchAllFromText,
  matchCode,
  matchLines,
} from './matching';

/** Build a tiny synthetic checklist from canonical codes for focused tests. */
function makeChecklist(codes: string[]): Checklist {
  const entries: ChecklistEntry[] = codes.map((code) => {
    const m = /^([A-Z]*)(\d+)$/.exec(code)!;
    return {
      code,
      display: m[1] ? `${m[1]} ${m[2]}` : code,
      teamCode: m[1] || 'FWC',
      teamName: m[1] || 'Especiais',
      number: parseInt(m[2], 10),
      type: 'player',
    };
  });
  return {
    entries,
    byCode: new Map(entries.map((e) => [e.code, e])),
    teams: [],
    total: entries.length,
  };
}

describe('extractCodes', () => {
  it('pulls a code out of noisy album text', () => {
    expect(extractCodes('FIFA WORLD CUP 2026 CIV 12')).toContain('CIV12');
  });

  it('handles codes with no space', () => {
    expect(extractCodes('panini EGY4 ©2026')).toContain('EGY4');
  });

  it('is case-insensitive and normalizes', () => {
    expect(extractCodes('civ 12')).toEqual(['CIV12']);
  });

  it('dedupes while preserving first-seen order', () => {
    expect(extractCodes('CIV 12 noise CIV12 more EGY4')).toEqual(['CIV12', 'EGY4']);
  });

  it('finds the standalone "00" logo but ignores stray single zeros', () => {
    expect(extractCodes('panini 00 logo')).toContain('00');
    // A lone "0" is OCR noise (a date, a jersey number), not the logo sticker.
    expect(extractCodes('only a 0 here')).toEqual([]);
  });

  it('does not absorb trailing letters of adjacent words into decoy codes', () => {
    // Regression: a greedy regex turned "WORLD CUP 2026" into "CUP202" and
    // "panini 00" into "NINI00", which could mis-correct to a real code.
    expect(extractCodes('FIFA WORLD CUP 2026')).not.toContain('CUP202');
    expect(extractCodes('panini 00 logo')).not.toContain('NINI00');
  });

  it('returns empty for text with no codes', () => {
    expect(extractCodes('just some words')).toEqual([]);
  });

  it('extracts the real codes from multi-line OCR (noise tokens allowed)', () => {
    const text = 'FIFA WORLD CUP 2026\nCIV 12\nMade in Italy\nFWC 1';
    const codes = extractCodes(text);
    // Surrounding text like "CUP 2026" can yield junk tokens; that's fine — the
    // real codes must be present and the noise gets dropped at match time.
    expect(codes).toContain('CIV12');
    expect(codes).toContain('FWC1');
  });

  it('extracts cleanly from text with no decoy numbers', () => {
    expect(extractCodes('back of sticker\nCIV 12\nFWC 1')).toEqual(['CIV12', 'FWC1']);
  });
});

describe('matchCode', () => {
  const list = makeChecklist(['CIV12', 'EGY4', 'FWC1', '00']);

  it('returns an exact match', () => {
    const r = matchCode('CIV12', list);
    expect(r.status).toBe('exact');
    expect(r.distance).toBe(0);
    expect(r.entry?.code).toBe('CIV12');
    expect(r.raw).toBe('CIV12');
  });

  it('auto-corrects ONLY the safe 0<->O confusion, never a blind letter substitution', () => {
    // 0<->O is the one same-length confusion safe to auto-correct (a digit/letter twin).
    const oList = makeChecklist(['POR10', 'CIV12']);
    expect(matchCode('P0R10', oList).entry?.code).toBe('POR10'); // zero read for the letter O
    // A blind single-letter substitution is NOT auto-corrected: it would manufacture a
    // DIFFERENT real code from a garbled read (the false positive a trading app must avoid).
    // Systematic letter confusions go through bestHighConfidenceConfusionMatchFromText instead.
    expect(matchCode('C1V12', list).status).toBe('unknown'); // '1'->'I' is not a safe blind edit
    expect(matchCode('GIV12', list).status).toBe('unknown'); // 'G'->'C' is not a safe blind edit
    expect(matchCode('EGYA', list).status).toBe('unknown'); // 'A'->'4' letter-for-digit, not safe
  });

  it('returns unknown when nothing is close enough', () => {
    const r = matchCode('ZZZ99', list);
    expect(r.status).toBe('unknown');
    expect(r.entry).toBeNull();
    expect(r.distance).toBe(-1);
    expect(r.raw).toBe('ZZZ99');
  });

  it('never blind-corrects a two-edit read', () => {
    expect(matchCode('CXX12', list, 1).status).toBe('unknown'); // 2 edits from CIV12
    expect(matchCode('CXX12', list, 2).status).toBe('unknown'); // still not a safe/curated edit
  });

  it('does not invent a code from a blind same-length substitution', () => {
    // "FWC12" is one blind digit substitution from FWC11 — NOT auto-corrected (would be a
    // wrong number, the worst kind of false positive). A miss is the safe outcome.
    const tie = makeChecklist(['FWC1', 'FWC11']);
    expect(matchCode('FWC12', tie).status).toBe('unknown');
  });

  it('rejects ambiguous, blind-substitution, and too-short reads', () => {
    // A blind letter substitution is never auto-corrected (GIV12 stays unknown — "G"->"C"
    // is not a safe blind edit), so a garbled read can't snap to a different real code.
    expect(matchCode('GIV12', checklist).status).toBe('unknown');
    // "EGYA" is one edit from EGY1..EGY9 — ambiguous and not a safe edit anyway.
    expect(matchCode('EGYA', checklist).status).toBe('unknown');
    // "SE3" (3 chars) is too short to correct safely.
    expect(matchCode('SE3', checklist).status).toBe('unknown');
  });

  it('restores a dropped THIN letter, but only when it is unambiguous', () => {
    // The OCR font reliably drops the thin "I": "CIV 12" reads as "CV 12". Because
    // the engine never drops a bold letter, the missing char can only be the thin
    // "I" — so "CV12" restores to CIV12 even though CPV12 also exists (reaching it
    // would require dropping a bold "P", which never happens).
    const thinOnly = makeChecklist(['CIV12', 'CPV12']);
    expect(matchCode('CV12', thinOnly).entry?.code).toBe('CIV12');
    // NOTE: against the real baked-in checklist, "CV12" is now ambiguous — the
    // Coca-Cola sticker "CC12" is a single substitution away (V→C), tying with the
    // thin-letter restore to CIV12 — so the real list returns unknown (never guesses).
    expect(matchCode('CV12', checklist).status).toBe('unknown');
    // A garbled-but-PRESENT middle char stays ambiguous: "C1V12" could be CIV12 or
    // CPV12 (we can't tell an "I" from a "P" in that slot).
    expect(matchCode('C1V12', checklist).status).toBe('unknown');
    // A missing BOLD letter is never restored (would be a wrong guess): "CPV12"
    // read as "CV12" is indistinguishable from CIV12, so we must not invent a "P".
    // (Implicit: only thin letters I/J/L/T are ever inserted.)
    const onlyBold = makeChecklist(['CPV12']);
    expect(matchCode('CV12', onlyBold).status).toBe('unknown');
  });

  it('works against the real baked-in checklist', () => {
    expect(matchCode('CIV12', checklist).status).toBe('exact');
    expect(matchCode('00', checklist).status).toBe('exact');
    // Thin-letter recovery still works on the real checklist: "BH12" restores the dropped I → BIH12
    // (unambiguous — no same-number code is a substitution away). "CV12" is NOT a good example
    // anymore: with Coca-Cola "CC12" in the list it's ambiguous (CIV12 via dropped-I vs CC12 via
    // C/V confusion), so the never-guess rule makes it unknown — see the thin-letter test above.
    expect(matchCode('BH12', checklist).entry?.code).toBe('BIH12');
    expect(matchCode('CV12', checklist).status).toBe('unknown');
    // A blind substitution is NEVER auto-corrected (no wrong guesses): GIV12/C1V12 stay unknown.
    expect(matchCode('GIV12', checklist).status).toBe('unknown');
    expect(matchCode('C1V12', checklist).status).toBe('unknown');
  });
});

describe('bestHighConfidenceConfusionMatchFromText (curated directed-confusion fallback)', () => {
  it('recovers a code through known directed glyph confusions, number preserved', () => {
    // "HSA 17" -> RSA17 (H read for R); "NJT 4" -> AUT4 (N for A, J for U). These systematic
    // font confusions are NOT reachable by a blind matchCode substitution — only the curated map.
    expect(bestHighConfidenceConfusionMatchFromText('HSA 17', checklist)?.entry?.code).toBe('RSA17');
    expect(bestHighConfidenceConfusionMatchFromText('NJT 4', checklist)?.entry?.code).toBe('AUT4');
  });

  it('never changes the number, and rejects unknown glyph swaps', () => {
    // "SWV 12" must NOT reach SWE12 (V->E is not a curated confusion) — the false positive guard.
    expect(bestHighConfidenceConfusionMatchFromText('SWV 12', checklist)).toBeNull();
    // No token / pure noise -> null.
    expect(bestHighConfidenceConfusionMatchFromText('just words', checklist)).toBeNull();
  });
});

describe('bestHighConfidenceExactAliasMatchFromText (per-sticker aliases, collision-guarded)', () => {
  it('applies a safe alias whose target is the strictly-nearest real code', () => {
    // "DXW4" is 2 edits from CUW4 and ≥3 from any other #4 code → safe.
    expect(bestHighConfidenceExactAliasMatchFromText('DXW4', checklist)?.entry?.code).toBe('CUW4');
  });

  it('REFUSES an alias that collides with another real same-number code (the 0-FP guard)', () => {
    // Each of these garbles is one edit from a DIFFERENT real code, so a different sticker could
    // produce it — mapping it would be a false positive. All must return null.
    expect(bestHighConfidenceExactAliasMatchFromText('NO2', checklist)).toBeNull(); // ~ real NOR2
    expect(bestHighConfidenceExactAliasMatchFromText('SN10', checklist)).toBeNull(); // ~ real SEN10
    expect(bestHighConfidenceExactAliasMatchFromText('TU10', checklist)).toBeNull(); // ~ real TUR10
    expect(bestHighConfidenceExactAliasMatchFromText('SWV12', checklist)).toBeNull(); // ~ real SWE12
    expect(bestHighConfidenceExactAliasMatchFromText('RO20', checklist)).toBeNull(); // ~ real CRO20
  });
});

describe('bestMatchFromText', () => {
  it('returns null when there are no codes', () => {
    expect(bestMatchFromText('no codes here', checklist)).toBeNull();
  });

  it('prefers an exact hit over other candidates', () => {
    const r = bestMatchFromText('garbage CIV12 EGY4', checklist);
    expect(r?.status).toBe('exact');
    expect(r?.entry?.code).toBe('CIV12');
  });

  it('falls back to a thin-letter correction', () => {
    const list = makeChecklist(['CIV12']);
    // 'CV12' is the dropped-thin-I reading of CIV12; matchCode restores it (the safe,
    // font-specific correction). Blind letter substitutions are no longer auto-corrected.
    const r = bestMatchFromText('noise CV12 more', list);
    expect(r?.status).toBe('corrected');
    expect(r?.entry?.code).toBe('CIV12');
  });

  it('returns unknown for the first token when nothing matches', () => {
    const list = makeChecklist(['CIV12']);
    const r = bestMatchFromText('ZZZ99 then WWW88', list);
    expect(r?.status).toBe('unknown');
    expect(r?.raw).toBe('ZZZ99');
  });

  it('reads a real multi-line sticker back', () => {
    const text = 'FIFA WORLD CUP 2026\nCIV 12\n© Panini';
    const r = bestMatchFromText(text, checklist);
    expect(r?.entry?.code).toBe('CIV12');
  });
});

describe('matchAllFromText', () => {
  it('resolves every distinct sticker when several are in view', () => {
    const text = 'FIFA WORLD CUP 2026 CIV 12\nEGY 4\nBRA 5\nFWC 1';
    const results = matchAllFromText(text, checklist);
    const codes = results.map((r) => r.entry?.code);
    expect(codes).toEqual(['CIV12', 'EGY4', 'BRA5', 'FWC1']);
    expect(results.every((r) => r.entry !== null)).toBe(true);
  });

  it('dedupes repeated codes and drops unmatched noise', () => {
    const text = 'CIV12 CIV 12 ZZZ99 EGY4';
    const codes = matchAllFromText(text, checklist).map((r) => r.entry?.code);
    expect(codes).toEqual(['CIV12', 'EGY4']);
  });

  it('returns an empty list when nothing matches', () => {
    expect(matchAllFromText('just some words', checklist)).toEqual([]);
  });
});

describe('matchLines', () => {
  it('matches one code per line from stacked crops', () => {
    const codes = matchLines('CIV 12\nEGY 4\nBRA 5', checklist).map((r) => r.entry?.code);
    expect(codes).toEqual(['CIV12', 'EGY4', 'BRA5']);
  });

  it('skips long lines so legal text cannot inject a code substring', () => {
    // "NID 9" sits inside a long legal-text line that the length gate drops.
    const text = 'CIV 12\nESTE CROMO E PARTE NID 9 INTEGRANTE DO ALBUM OFICIAL';
    expect(matchLines(text, checklist).map((r) => r.entry?.code)).toEqual(['CIV12']);
  });

  it('dedupes repeated codes and ignores blank/unknown lines', () => {
    expect(matchLines('EGY 4\n\nEGY 4\nZZZ 99', checklist).map((r) => r.entry?.code)).toEqual([
      'EGY4',
    ]);
  });

  it('tolerates a little noise on a short code line', () => {
    expect(matchLines('EGY 4 7', checklist).map((r) => r.entry?.code)).toEqual(['EGY4']);
  });

  it('corrects thin-letter slips on individual codes in a group', () => {
    const text = 'BH12 EGY4'; // "BH12" is the dropped-thin-I reading of BIH12 (unambiguous)
    const codes = matchAllFromText(text, checklist).map((r) => r.entry?.code);
    expect(codes).toContain('BIH12');
    expect(codes).toContain('EGY4');
  });
});

File: android/app/src/test/java/br/com/fiquemsabendo/figurinhas/domain/MatchingTest.kt
package br.com.fiquemsabendo.figurinhas.domain

import br.com.fiquemsabendo.figurinhas.data.checklist
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull
import kotlin.test.assertTrue

class MatchingTest {
    /** Tiny synthetic checklist from canonical codes for focused tests. */
    private fun makeChecklist(codes: List<String>): Checklist {
        val re = Regex("^([A-Z]*)(\\d+)$")
        val entries = codes.map { code ->
            val m = re.matchEntire(code)!!
            ChecklistEntry(
                code = code,
                display = if (m.groupValues[1].isNotEmpty()) "${m.groupValues[1]} ${m.groupValues[2]}" else code,
                teamCode = m.groupValues[1].ifEmpty { "FWC" },
                teamName = m.groupValues[1].ifEmpty { "Especiais" },
                number = m.groupValues[2].toInt(),
                type = StickerType.PLAYER,
            )
        }
        return Checklist(entries, entries.associateBy { it.code }, emptyList(), entries.size)
    }

    // ---- extractCodes ----
    @Test fun extract_pulls_code_from_noisy_text() {
        assertTrue("CIV12" in extractCodes("FIFA WORLD CUP 2026 CIV 12"))
    }
    @Test fun extract_no_space() { assertTrue("EGY4" in extractCodes("panini EGY4 ©2026")) }
    @Test fun extract_case_insensitive() { assertEquals(listOf("CIV12"), extractCodes("civ 12")) }
    @Test fun extract_dedupes_first_seen() {
        assertEquals(listOf("CIV12", "EGY4"), extractCodes("CIV 12 noise CIV12 more EGY4"))
    }
    @Test fun extract_logo_but_not_stray_zero() {
        assertTrue("00" in extractCodes("panini 00 logo"))
        assertEquals(emptyList(), extractCodes("only a 0 here"))
    }
    @Test fun extract_no_decoy_from_adjacent_words() {
        assertTrue("CUP202" !in extractCodes("FIFA WORLD CUP 2026"))
        assertTrue("NINI00" !in extractCodes("panini 00 logo"))
    }
    @Test fun extract_empty_when_none() { assertEquals(emptyList(), extractCodes("just some words")) }
    @Test fun extract_multiline_noise_allowed() {
        val codes = extractCodes("FIFA WORLD CUP 2026\nCIV 12\nMade in Italy\nFWC 1")
        assertTrue("CIV12" in codes); assertTrue("FWC1" in codes)
    }
    @Test fun extract_clean_no_decoys() {
        assertEquals(listOf("CIV12", "FWC1"), extractCodes("back of sticker\nCIV 12\nFWC 1"))
    }

    // ---- matchCode ----
    private val list = makeChecklist(listOf("CIV12", "EGY4", "FWC1", "00"))

    @Test fun match_exact() {
        val r = matchCode("CIV12", list)
        assertEquals(MatchStatus.EXACT, r.status); assertEquals(0, r.distance)
        assertEquals("CIV12", r.entry?.code); assertEquals("CIV12", r.raw)
    }
    @Test fun match_rejects_unsafe_same_length_substitution() {
        val r = matchCode("C1V12", list)
        assertEquals(MatchStatus.UNKNOWN, r.status); assertNull(r.entry); assertEquals(-1, r.distance)
    }
    @Test fun match_rejects_unsafe_letter_for_digit() {
        val r = matchCode("EGYA", list) // 'A' read instead of '4'
        assertEquals(MatchStatus.UNKNOWN, r.status); assertNull(r.entry); assertEquals(-1, r.distance)
    }
    @Test fun match_unknown_when_far() {
        val r = matchCode("ZZZ99", list)
        assertEquals(MatchStatus.UNKNOWN, r.status); assertNull(r.entry); assertEquals(-1, r.distance); assertEquals("ZZZ99", r.raw)
    }
    @Test fun match_respects_maxDistance() {
        val safe = makeChecklist(listOf("OO12"))
        assertEquals(MatchStatus.UNKNOWN, matchCode("0O12", safe, 0).status)
        val r = matchCode("0O12", safe, 1)
        assertEquals(MatchStatus.CORRECTED, r.status); assertEquals("OO12", r.entry?.code); assertEquals(1, r.distance)
    }
    @Test fun match_does_not_use_unsafe_equal_length_tie() {
        val tie = makeChecklist(listOf("FWC1", "FWC11"))
        assertEquals(MatchStatus.UNKNOWN, matchCode("FWC12", tie, 1).status)
    }
    @Test fun match_unique_or_reject() {
        assertEquals(MatchStatus.UNKNOWN, matchCode("GIV12", checklist).status)
        assertEquals(MatchStatus.UNKNOWN, matchCode("EGYA", checklist).status)
        assertEquals(MatchStatus.UNKNOWN, matchCode("SE3", checklist).status)
    }
    @Test fun match_rejects_known_false_positive_patterns() {
        val unsafeCorrections = makeChecklist(listOf("HAI2", "RSA6", "EGY5"))
        assertEquals(MatchStatus.UNKNOWN, matchCode("WAI2", unsafeCorrections).status)
        assertEquals(MatchStatus.UNKNOWN, matchCode("RSA8", unsafeCorrections).status)
        assertEquals(MatchStatus.UNKNOWN, matchCode("EGY8", unsafeCorrections).status)
    }
    @Test fun match_restores_thin_letter_only_when_unambiguous() {
        assertEquals("CIV12", matchCode("CV12", checklist).entry?.code)
        assertEquals("CIV4", matchCode("CV4", checklist).entry?.code)
        assertEquals("BIH12", matchCode("BH12", checklist).entry?.code)
        assertEquals(MatchStatus.UNKNOWN, matchCode("C1V12", checklist).status)
        assertEquals(MatchStatus.UNKNOWN, matchCode("C1V4", checklist).status)
        val onlyBold = makeChecklist(listOf("CPV12"))
        assertEquals(MatchStatus.UNKNOWN, matchCode("CV12", onlyBold).status)
    }
    @Test fun match_rejects_edge_thin_letter_restoration() {
        assertEquals(MatchStatus.UNKNOWN, matchCode("RN10", checklist).status)
        assertEquals(MatchStatus.UNKNOWN, matchCode("AU4", checklist).status)
    }
    @Test fun match_real_checklist() {
        assertEquals(MatchStatus.EXACT, matchCode("CIV12", checklist).status)
        assertEquals(MatchStatus.EXACT, matchCode("00", checklist).status)
        assertEquals(MatchStatus.UNKNOWN, matchCode("GIV12", checklist).status)
        assertEquals(MatchStatus.UNKNOWN, matchCode("C1V12", checklist).status)
    }

    // ---- bestMatchFromText ----
    @Test fun best_null_when_no_codes() { assertNull(bestMatchFromText("no codes here", checklist)) }
    @Test fun best_prefers_exact() {
        val r = bestMatchFromText("garbage CIV12 EGY4", checklist)
        assertEquals(MatchStatus.EXACT, r?.status); assertEquals("CIV12", r?.entry?.code)
    }
    @Test fun best_falls_back_to_conservative_thin_correction() {
        val l = makeChecklist(listOf("CIV12"))
        val r = bestMatchFromText("noise CV12 more", l)
        assertEquals(MatchStatus.CORRECTED, r?.status); assertEquals("CIV12", r?.entry?.code)
    }
    @Test fun best_unknown_first_token() {
        val l = makeChecklist(listOf("CIV12"))
        val r = bestMatchFromText("ZZZ99 then WWW88", l)
        assertEquals(MatchStatus.UNKNOWN, r?.status); assertEquals("ZZZ99", r?.raw)
    }
    @Test fun best_reads_multiline_back() {
        val r = bestMatchFromText("FIFA WORLD CUP 2026\nCIV 12\n© Panini", checklist)
        assertEquals("CIV12", r?.entry?.code)
    }
    @Test fun high_confidence_confusion_recovers_only_known_letter_confusions_with_exact_digits() {
        val r = bestHighConfidenceConfusionMatchFromText("NJT 4", checklist)
        assertEquals("AUT4", r?.entry?.code)
        assertEquals(2, r?.distance)
        assertNull(bestMatchFromText("NJT 4", checklist)?.entry)
    }
    @Test fun high_confidence_confusion_recovers_verified_merged_m_shape() {
        val nShape = bestHighConfidenceConfusionMatchFromText("NEX 15", checklist)
        assertEquals("MEX15", nShape?.entry?.code)
        assertEquals(1, nShape?.distance)
        val iShape = bestHighConfidenceConfusionMatchFromText("IEX 15", checklist)
        assertEquals("MEX15", iShape?.entry?.code)
        assertEquals(1, iShape?.distance)
        assertNull(bestHighConfidenceConfusionMatchFromText("NEX 16", makeChecklist(listOf("MEX15"))))
    }
    @Test fun high_confidence_confusion_recovers_verified_irq20_shape() {
        val r = bestHighConfidenceConfusionMatchFromText("IWJ 20", checklist)
        assertEquals("IRQ20", r?.entry?.code)
        assertEquals(2, r?.distance)
    }
    @Test fun high_confidence_confusion_recovers_verified_tun10_shape() {
        val r = bestHighConfidenceConfusionMatchFromText("TIN 10", checklist)
        assertEquals("TUN10", r?.entry?.code)
        assertEquals(1, r?.distance)
        val uShape = bestHighConfidenceConfusionMatchFromText("UIN 10", checklist)
        assertEquals("TUN10", uShape?.entry?.code)
        assertEquals(2, uShape?.distance)
    }
    @Test fun high_confidence_confusion_recovers_verified_egy4_shape() {
        val r = bestHighConfidenceConfusionMatchFromText("ECY 4", checklist)
        assertEquals("EGY4", r?.entry?.code)
        assertEquals(1, r?.distance)
        val fShape = bestHighConfidenceConfusionMatchFromText("FGY 4", checklist)
        assertEquals("EGY4", fShape?.entry?.code)
        assertEquals(1, fShape?.distance)
    }
    @Test fun high_confidence_confusion_recovers_verified_aut_shapes() {
        val mShape = bestHighConfidenceConfusionMatchFromText("MIT 4", checklist)
        assertEquals("AUT4", mShape?.entry?.code)
        assertEquals(2, mShape?.distance)
        val wShape = bestHighConfidenceConfusionMatchFromText("WJT 4", checklist)
        assertEquals("AUT4", wShape?.entry?.code)
        assertEquals(2, wShape?.distance)
    }
    @Test fun high_confidence_exact_alias_recovers_verified_cuw4_shapes() {
        assertEquals("CUW4", bestHighConfidenceExactAliasMatchFromText("DXW 4", checklist)?.entry?.code)
        assertEquals("CUW4", bestHighConfidenceExactAliasMatchFromText("DAV 4", checklist)?.entry?.code)
        assertNull(bestHighConfidenceExactAliasMatchFromText("DXW 5", checklist))
    }
    @Test fun high_confidence_confusion_recovers_verified_pixel_shapes() {
        val verifiedPixelShapes = linkedMapOf(
            "DUA 19" to "GHA19",
            "ER 4" to "GER4",
            "HSA 17" to "RSA17",
            "RSM 17" to "RSA17",
            "RGA 17" to "RSA17",
            "NEN 20" to "NOR20",
            "EN 20" to "NOR20",
            "MN 20" to "NOR20",
            "CWV 12" to "CIV12",
            "OV 4" to "CIV4",
            "SXV 4" to "CIV4",
            "CNV 4" to "CIV4",
            "NB 18" to "NZL18",
            "NO 2" to "AUS2",
            "WAI 2" to "AUS2",
            "WIU 8" to "AUT8",
            "NGA 6" to "RSA6",
            "NEA 6" to "RSA6",
            "FGA 19" to "RSA19",
            "GIE 8" to "SWE8",
            "SWT 8" to "SWE8",
            "EX 14" to "SUI14",
            "WX 14" to "SUI14",
            "UJMJ 10" to "TUN10",
            "IL 10" to "TUN10",
            "EG 1" to "ALG1",
            "MB 1" to "ALG1",
            "3AT 17" to "QAT17",
            "OAT 17" to "QAT17",
            "SOT 17" to "QAT17",
            "TOT 17" to "QAT17",
            "COT 17" to "QAT17",
            "SBT 17" to "QAT17",
            "SWJ 10" to "IRN10",
            "OWJ 10" to "IRN10",
            "MN 10" to "IRN10",
            "GN 10" to "IRN10",
            "ON 15" to "IRN15",
            "SVU 15" to "IRN15",
            "MN 15" to "IRN15",
            "DH 12" to "BIH12",
            "RH 12" to "BIH12",
            "SOU 12" to "BIH12",
            "OJW 4" to "CUW4",
            "DAY 4" to "CUW4",
            "OAV 4" to "CUW4",
            "DXW 4" to "CUW4",
            "DAV 4" to "CUW4",
            "BN 10" to "POR10",
            "TXN 10" to "POR10",
            "IWJ 20" to "IRQ20",
            "SE 20" to "IRQ20",
            "SXJ 20" to "IRQ20",
            "WO 20" to "IRQ20",
        )
        for ((read, expected) in verifiedPixelShapes) {
            assertEquals(expected, bestHighConfidenceConfusionMatchFromText(read, checklist)?.entry?.code, read)
        }
    }
    @Test fun high_confidence_confusion_rejects_digit_changes_and_unknown_letter_pairs() {
        val onlyAut4 = makeChecklist(listOf("AUT4"))
        assertNull(bestHighConfidenceConfusionMatchFromText("NJT 5", onlyAut4))
        assertNull(bestHighConfidenceConfusionMatchFromText("NRT 4", checklist))
        assertNull(bestHighConfidenceConfusionMatchFromText("RGA 8", checklist))
        assertNull(bestHighConfidenceConfusionMatchFromText("SXJ 18", makeChecklist(listOf("IRQ18"))))
        assertNull(bestHighConfidenceConfusionMatchFromText("SWJ 10", makeChecklist(listOf("IRQ10"))))
    }

    // ---- matchAllFromText ----
    @Test fun all_resolves_each_distinct() {
        val results = matchAllFromText("FIFA WORLD CUP 2026 CIV 12\nEGY 4\nBRA 5\nFWC 1", checklist)
        assertEquals(listOf("CIV12", "EGY4", "BRA5", "FWC1"), results.map { it.entry?.code })
        assertTrue(results.all { it.entry != null })
    }
    @Test fun all_dedupes_and_drops_noise() {
        assertEquals(listOf("CIV12", "EGY4"), matchAllFromText("CIV12 CIV 12 ZZZ99 EGY4", checklist).map { it.entry?.code })
    }
    @Test fun all_empty_when_none() { assertEquals(emptyList(), matchAllFromText("just some words", checklist)) }

    // ---- matchLines ----
    @Test fun lines_one_per_line() {
        assertEquals(listOf("CIV12", "EGY4", "BRA5"), matchLines("CIV 12\nEGY 4\nBRA 5", checklist).map { it.entry?.code })
    }
    @Test fun lines_skips_long_legal_text() {
        val text = "CIV 12\nESTE CROMO E PARTE NID 9 INTEGRANTE DO ALBUM OFICIAL"
        assertEquals(listOf("CIV12"), matchLines(text, checklist).map { it.entry?.code })
    }
    @Test fun lines_dedupes_and_ignores_blank_unknown() {
        assertEquals(listOf("EGY4"), matchLines("EGY 4\n\nEGY 4\nZZZ 99", checklist).map { it.entry?.code })
    }
    @Test fun lines_tolerates_noise_on_short_line() {
        assertEquals(listOf("EGY4"), matchLines("EGY 4 7", checklist).map { it.entry?.code })
    }
}
